man/man1/cs2cs.1
.\" release 4
.nr LL 5.5i
.ad b
.hy 1
.TH CS2CS 1 "2000/03/21 Rel. 4.4"
.SH NAME
cs2cs \- cartographic coordinate system filter
.SH SYNOPSIS
.B cs2cs
[
.B \-eEfIlrstvwW
[
.I args
] ] [
.B +opts[=arg]
]
.br
      [ \fB+to\fR [\fB+opts\fR[\fB=arg\fR]] ]
file[s]
.SH DESCRIPTION
.I Cs2cs
performs transformation between the source and destination cartographic
coordinate system on a set of input points.  The coordinate system
transformation can include translation between projected and geographic
coordinates as well as the application of datum shifts.
.PP
The following control parameters can appear in any order:
.TP
.BI \-I
method to specify inverse translation, convert from \fB+to\fR coordinate
system to the primary coordinate system defined.
.TP
.BI \-t "a"
.I A
specifies a character employed as the first character to denote
a control line to be passed through without processing.
This option applicable to ascii input only.
(# is the default value).
.TP
.BI \-e " string"
.I String
is an arbitrary string to be output if an error is detected during
data transformations.
The default value is: *\et*.
Note that if the
.B \-b,
.B \-i
or
.B \-o
options are employed, an error is returned as HUGE_VAL
value for both return values.
.TP
.BI \-E
causes the input coordinates to be copied to the output line
prior to printing the converted values.
.TP
.BI \-l "[p|P|=|e|u|d]" id
List projection identifiers with
.B \-l,
.B \-lp
or
.B \-lP (expanded)
that can be selected with
.B +proj.
.BI \-l= id
gives expanded description of projection
.I id.
List ellipsoid identifiers with
.B \-le,
that can be selected with
.B +ellps,
.B \-lu
list of cartesian to meter conversion factors
that can be selected with
.B +units
or
.B \-ld
list of datums that can be selected with
.B +datum.
.TP
.BI \-r
This options reverses the order of the
expected input from longitude-latitude or x-y to latitude-longitude or y-x.
.TP
.BI \-s
This options reverses the order of the
output from x-y or longitude-latitude to y-x or latitude-longitude.
.TP
.BI \-f " format"
.I Format
is a
.I printf
format string to control the form of the output values.
For inverse projections, the output will be in degrees when this option
is employed.
If a format is specified for inverse projection the
output data will be in decimal degrees.
The default format is "%.2f" for forward projection and DMS
for inverse.
.TP
.BI \-[w|W] n
.I N
is the number of significant fractional digits to employ for
seconds output (when the option is not specified,
.B \-w3
is assumed).
When
.B \-W
is employed the fields will be constant width and with leading zeroes.
.TP
.B \-v
causes a listing of cartographic control parameters tested for and
used by the program to be printed prior to input data.
.PP
The
.B +args
run-line arguments are associated with cartographic parameters
and usage varies with projection and for a complete description see
.I "Cartographic Projection Procedures for the UNIX Environment\(emA User's Manual"
and supplementary documentation for Release 4.
.PP
The \fIcs2cs\fR program requires two coordinate system definitions.  The
first (or primary is defined based on all projection parameters not
appearing after the \fB+to\fR argument.  All projection parameters
appearing after the \fB+to\fR argument are considered the definition
of the second coordinate system.  If there is no second coordinate system
defined, a geographic coordinate system based on the datum and ellipsoid of
the source coordinate system is assumed.  Note that the source and destination
coordinate system can both be projections, both be geographic, or one of
each and may have the same or different datums.
.PP
Additional projection control parameters may be contained in two
auxiliary control files:
the first is optionally referenced with the
.BI +init= file:id
and the second is always processed after the name
of the projection has been established from either the run-line
or the contents of
.B +init
file.
The environment parameter
.B PROJ_LIB
establishes the default directory for a file reference without
an absolute path.  This is also used for supporting files like
datum shift files.
.PP
One or more
.I files
(processed in left to right order)
specify the source of data to be transformed.
A \- will specify the location of processing standard input.
If no files are specified, the input is assumed to be from
.I stdin.
For input data the two data values must be in the
first two white space separated fields and
when both input and output are ASCII all trailing portions
of the input line are appended to the output line.
.PP
Input geographic data
(longitude and latitude) must be in DMS or decimal degrees format and input
cartesian data must be in units consistent with the ellipsoid
major axis or sphere radius units.
Output geographic coordinates will normally be in DMS format (use
.B \-f %.12f
for decimal degrees with 12 decimal places), while
projected (cartesian) coordinates will be in linear (meter, feet) units.

.SH EXAMPLE
The following script
.RS 5
 \f(CWcs2cs +proj=latlong +datum=NAD83
       +to +proj=utm +zone=10 +datum=NAD27 \-r <<EOF
 45d15'33.1"	111.5W
 45d15.551666667N	\-111d30
 +45.25919444444	111d30'000w
 EOF\fR
.RE
will transform the input NAD83 geographic coordinates into NAD27 coordinates
in the UTM projection with zone 10 selected.
The geographic values of this example are equivalent and meant
as examples of various forms of DMS input.
The x\-y output data will appear as three lines of:
.RS 5
 \f(CW1402285.99      5076292.42 0.000
.RE
.SH SEE ALSO
.B proj(1),
.br
.I "Cartographic Projection Procedures for the UNIX Environment\(emA User's Manual,"
(Evenden, 1990, Open-file report 90\-284).
.br
.I "Map Projections Used by the U. S. Geological Survey"
(Snyder, 1984,
USGS Bulletin 1532).
.br
.I "Map Projections\(emA Working Manual"
(Snyder, 1988, USGS Prof. Paper 1395).
.br
.I "An Album of Map Projections"
(Snyder & Voxland, 1989, USGS Prof. Paper 1453).
.SH BUGS
A list of known bugs can found at https://github.com/OSGeo/proj.4/issues
where new bug reports can be submitted too.
.SH HOME PAGE
http://proj4.org/


NEWS
= GDAL/OGR 1.11.3 Release Notes = 

The 1.11.3 release is a bug fix release.

== Build (Unix) ==
 * Support custom buildflags in GRASS plugin (#5850)
 * GRASS 7.0.0 support (#5852)
 * fix detection of ECW SDK 5.1 on Mac (#5867, #5390)
 * Add support for Poppler >= 0.31.0 (#6118)

== Port ==
 * /vsitar/: remove useless validation test that prevents from opening valid .tar files (#5864)

== Algorithms ==
 * gdal_grid: fix crash in optimized mode with GCC 4.4 on 64bit (#5987)
 * Fix failure in GDALSuggestedWarpOut2() when computing output image size in case the top-left and bottom-right corners tranform to the same point (#5980)
 * Rasterize: add check to avoid burning negative x values (#5641)

== GDAL core ==
 * GMLJP2 reader: add compatibility with OGC CRS URL as found in GMLJP2v2 (#5940)
 * GMLJP2: on reading, don't do axis inversation if there's an explicit axisName requesting easting, northing order (#5960)
 * GMLJP2: add missing rangeParameters element to validate against GMLJP2 schema (#5707)

== Utilities ==
 * gdalbuildvrt: fix potential crash when using -b switch (#6095)
 * gdalserver: fix compilation with recent GNU libc (such as in Arch Linux) (#6073)
 * gdalwarp: emit error message if file specified with -cutline cannot be opened (#5921)
 * gdalwarp: fix 1.11.2 regression when invoking several times gdalwarp with several input files on the same target file, and when the input files have a nodata setting: only the last input file was warped onto the target dataset (#5909)
 * ogr2ogr: fix crash with -clipdst when a reprojection fails before (#5973)
 * ogr_layer_algebra.py: for Update, Clip and Erase, only creates attribute of input layer by default (#5976)

== GDAL drivers ==

ECRG driver:
 * change subdataset definition to make sure that they only consist of frames of same scale (#6043)

ERS driver:
 * fix SetProjection() (#5840)

GeoRaster driver:
 * New SRID search (#5911)
 * Fix user-defined SRID issue (#5881)

GTiff driver:
 * fix GTiffDataset::IsBlockAvailable() wrong behaviour when compiling against internal libtiff, when a BigTIFF file has a TileByteCounts with LONG/4-byte counts and not LONG8 (#6001)
 * avoid generated corrupted  right-most and bottom-most tiles for 12-bit JPEG-compressed (#5971)
 * on reading attach PROJ4 extension node for Mercator_Auxiliary_Sphere / WebMercator (#5924)

HDF4 driver:
 * fix AnyTypeToDouble() to use proper type (int instead of long) to work with DFNT_INT32/DFNT_UINT32 on 64-bit Linux (#5965)

HF2 driver:
 * fix reading side of the driver to work on architectures where char is unsigned, like PPC or ARM (#6082)

NetCDF driver:
 * fix crash on opening a NOAA dataset (#5962)
 * fix computation of inverse flattening (#5858)

Northwoord driver:
 * Fix computation of intermediate color table values on non-Intel platforms (#6091)
 * NWT_GRD: don't advertize scale/offset as they are transparently applied in IReadBlock() (#5839).

RPFTOC driver:
 * add tweak for weird relative directory names in the A.TOC file (#5979)

TIL driver:
 * fix half pixel shift in geo registration (#5961)

VRT driver:
 * add more checks to CheckCompatibleForDatasetIO() to avoid issues with overview bands (#5954)
 * fix rounding of output window size on VRTSimpleSource (#5874)

== OGRSpatialReference ==
 * morphFromESRI(): map Mercator_Auxiliary_Sphere to EPSG:3857 (#5924)

== OGR drivers ==

EDIGEO driver:
 * backport conversion of atof() to CPLAtof() to avoid truncation of floating point numbers including coordinates on a non C locale (#6070)

GML driver:
 * accept choice between gml:polygonProperty and gml:multiPolygonProperty (#5977)
 * fix GML_ATTRIBUTES_TO_OGR_FIELDS=YES to work correctly with xlink:href too (#5970); and also backport a trunk fix for writing the correctly path to attributes in nested elements
 * VFR: add support for OriginalniHraniceOmpv (#5853)

GeoJSON driver:
 * avoid truncation of real numbers on reading (#5882)
 * TopoJSON: fix segfault when reading https://raw.githubusercontent.com/mbostock/topojson/master/test/topojson/properties-id-computed.json (#5919)
 * GeoJSON writer: make string comparison for authority name case insensitive so as to recognize lowercase 'epsg' (#4995)

GPKG driver:
 * on reading, recognize GEOMCOLLECTION to be compatible of GDAL 2.0. Still writing GEOMETRYCOLLECTION in 1.11 branch (#5937)
 * escape column names on reading/insert/update (#5879)

MSSQLSpatial driver:
 * Add support for creating features with default values (#5957)

OpenFileGDB driver:
 * fix 1.11.2 regression with FileGDB v9 where the presence of a non spatial table can cause other layers not to be listed (#5875)

PG/PGDump drivers:
 * fix truncation of fields to work with multi-byte UTF-8 characters (#5854)

Shape driver:
 * SHPReadOGRFeature() should free the passed psShape when not NULL and if the DBF record is deleted, so as to avoid later reads to faild (linked to #5974)

SQLite driver:
 * Fix issue with consecutive identical characters in layer name with ExecuteSQL() (#6107)

== SWIG Language Bindings ==

Perl bindings:
 * Fix creation of Field Definition object (#6099)

Python bindings:
 * Fix build with SWIG 3.0.6 (#6045)

= GDAL/OGR 1.11.2 Release Notes = 

The 1.11.2 release is a bug fix release.

== Build ==

  * configure: fix detection of OCI by changing linking order to please modern GCC (#5550)
  * configure: fix test to accept MariaDB 10.X as valid MySQL (#5722)
  * More compiler warnings addressed (#5414)
  * Windows build of PDF: fix compilation issue with Visual Studio 2012 (#5744)
  * Windows build: Add support for MrSID 9.1 SDK (#5814)
  * Windows build: when building netCDF, HDF4, HDF5 as plugins, call registration of 'sub-drivers' GMT, HDF4Image and HDF5Image (#5802)

== Port ==

 * CSV finder: Stop probing for csv/horiz_cs.csv. (#5698)
 * /vsicurl/: avoid reading after end-of-file and fix failure when reading more than 16MB in a single time (#5786)
 * CPLHexToBinary(): faster implementation (#5812)


proj4/src/cs2cs.c
/******************************************************************************
 * Project:  PROJ.4
 * Purpose:  Mainline program sort of like ``proj'' for converting between
 *           two coordinate systems.
 * Author:   Frank Warmerdam, warmerda@home.com
 *
 ******************************************************************************
 * Copyright (c) 2000, Frank Warmerdam
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#include "projects.h"
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <math.h>
#include "emess.h"
#include <locale.h>

#define MAX_LINE 1000
#define MAX_PARGS 100

static projPJ   fromProj, toProj;

static int
reversein = 0,	/* != 0 reverse input arguments */
reverseout = 0,	/* != 0 reverse output arguments */
echoin = 0,	/* echo input data to output line */
tag = '#';	/* beginning of line tag character */
	static char
*oform = (char *)0,	/* output format for x-y or decimal degrees */
*oterr = "*\t*",	/* output line for unprojectable input */
*usage =
"%s\nusage: %s [ -eEfIlrstvwW [args] ] [ +opts[=arg] ]\n"
"                   [+to [+opts[=arg] [ files ]\n";

static double (*informat)(const char *, 
                          char **); /* input data deformatter function */


/************************************************************************/
/*                              process()                               */
/*                                                                      */
/*      File processing function.                                       */
/************************************************************************/
static void process(FILE *fid) 

{
    char line[MAX_LINE+3], *s, pline[40];
    projUV data;

    for (;;) {
        double z;

        ++emess_dat.File_line;
        if (!(s = fgets(line, MAX_LINE, fid)))
            break;
        if (!strchr(s, '\n')) { /* overlong line */
            int c;
            (void)strcat(s, "\n");
				/* gobble up to newline */
            while ((c = fgetc(fid)) != EOF && c != '\n') ;
        }
        if (*s == tag) {
            fputs(line, stdout);
            continue;
        }

        if (reversein) {
            data.v = (*informat)(s, &s);
            data.u = (*informat)(s, &s);
        } else {
            data.u = (*informat)(s, &s);
            data.v = (*informat)(s, &s);
        }

        z = strtod( s, &s );

        if (data.v == HUGE_VAL)
            data.u = HUGE_VAL;

        if (!*s && (s > line)) --s; /* assumed we gobbled \n */

        if ( echoin) {
            int t;
            t = *s;
            *s = '\0';
            (void)fputs(line, stdout);
            *s = t;
            putchar('\t');
        }

        if (data.u != HUGE_VAL) {
            if( pj_transform( fromProj, toProj, 1, 0, 
                              &(data.u), &(data.v), &z ) != 0 )
            {
                data.u = HUGE_VAL;
                data.v = HUGE_VAL;
                emess(-3,"pj_transform(): %s", pj_strerrno(pj_errno));
            }
        }

        if (data.u == HUGE_VAL) /* error output */
            fputs(oterr, stdout);

        else if (pj_is_latlong(toProj) && !oform) {	/*ascii DMS output */
            if (reverseout) {
                fputs(rtodms(pline, data.v, 'N', 'S'), stdout);
                putchar('\t');
                fputs(rtodms(pline, data.u, 'E', 'W'), stdout);
            } else {
                fputs(rtodms(pline, data.u, 'E', 'W'), stdout);
                putchar('\t');
                fputs(rtodms(pline, data.v, 'N', 'S'), stdout);
            }

        } else {	/* x-y or decimal degree ascii output */
            if ( pj_is_latlong(toProj) ) {
                data.v *= RAD_TO_DEG;
                data.u *= RAD_TO_DEG;
            }
            if (reverseout) {
                printf(oform,data.v); putchar('\t');
                printf(oform,data.u);
            } else {
                printf(oform,data.u); putchar('\t');
                printf(oform,data.v);
            }
        }

        putchar(' ');
        if( oform != NULL )
            printf( oform, z );
        else
            printf( "%.3f", z );
        if( s )
            printf( "%s", s );
        else
            printf( "\n" );
    }
}

/************************************************************************/
/*                                main()                                */
/************************************************************************/

int main(int argc, char **argv) 
{
    char *arg, **eargv = argv, *from_argv[MAX_PARGS], *to_argv[MAX_PARGS];
    FILE *fid;
    int from_argc=0, to_argc=0, eargc = 0, c, mon = 0;
    int have_to_flag = 0, inverse = 0, i;
    int use_env_locale = 0;

    /* This is just to check that pj_init() is locale-safe */
    /* Used by nad/testvarious */
    if( getenv("PROJ_USE_ENV_LOCALE") != NULL )
        use_env_locale = 1;

    if ((emess_dat.Prog_name = strrchr(*argv,DIR_CHAR)) != NULL)
        ++emess_dat.Prog_name;
    else emess_dat.Prog_name = *argv;
    inverse = ! strncmp(emess_dat.Prog_name, "inv", 3);
    if (argc <= 1 ) {
        (void)fprintf(stderr, usage, pj_get_release(), emess_dat.Prog_name);
        exit (0);
    }
    /* process run line arguments */
    while (--argc > 0) { /* collect run line arguments */
        if(**++argv == '-') for(arg = *argv;;) {
            switch(*++arg) {
              case '\0': /* position of "stdin" */
                if (arg[-1] == '-') eargv[eargc++] = "-";
                break;
              case 'v': /* monitor dump of initialization */
                mon = 1;
                continue;
              case 'I': /* alt. method to spec inverse */
                inverse = 1;
                continue;
              case 'E': /* echo ascii input to ascii output */
                echoin = 1;
                continue;
              case 't': /* set col. one char */
                if (arg[1]) tag = *++arg;
                else emess(1,"missing -t col. 1 tag");
                continue;
              case 'l': /* list projections, ellipses or units */
                if (!arg[1] || arg[1] == 'p' || arg[1] == 'P') {
                    /* list projections */
                    struct PJ_LIST *lp;
                    int do_long = arg[1] == 'P', c;
                    char *str;

                    for (lp = pj_get_list_ref() ; lp->id ; ++lp) {
                        (void)printf("%s : ", lp->id);
                        if (do_long)  /* possibly multiline description */
                            (void)puts(*lp->descr);
                        else { /* first line, only */
                            str = *lp->descr;
                            while ((c = *str++) && c != '\n')
                                putchar(c);
                            putchar('\n');
                        }
                    }
                } else if (arg[1] == '=') { /* list projection 'descr' */
                    struct PJ_LIST *lp;

                    arg += 2;
                    for (lp = pj_get_list_ref() ; lp->id ; ++lp)
                        if (!strcmp(lp->id, arg)) {
                            (void)printf("%9s : %s\n", lp->id, *lp->descr);
                            break;
                        }
                } else if (arg[1] == 'e') { /* list ellipses */
                    struct PJ_ELLPS *le;

                    for (le = pj_get_ellps_ref(); le->id ; ++le)
                        (void)printf("%9s %-16s %-16s %s\n",
                                     le->id, le->major, le->ell, le->name);
                } else if (arg[1] == 'u') { /* list units */
                    struct PJ_UNITS *lu;

                    for (lu = pj_get_units_ref(); lu->id ; ++lu)
                        (void)printf("%12s %-20s %s\n",
                                     lu->id, lu->to_meter, lu->name);
                } else if (arg[1] == 'd') { /* list datums */
                    struct PJ_DATUMS *ld;

                    printf("__datum_id__ __ellipse___ __definition/comments______________________________\n" );
                    for (ld = pj_get_datums_ref(); ld->id ; ++ld)
                    {
                        printf("%12s %-12s %-30s\n",
                               ld->id, ld->ellipse_id, ld->defn);
                        if( ld->comments != NULL && strlen(ld->comments) > 0 )
                            printf( "%25s %s\n", " ", ld->comments );
                    }
                } else if( arg[1] == 'm') { /* list prime meridians */
                    struct PJ_PRIME_MERIDIANS *lpm;

                    for (lpm = pj_get_prime_meridians_ref(); lpm->id ; ++lpm)
                        (void)printf("%12s %-30s\n",
                                     lpm->id, lpm->defn);
                } else
                    emess(1,"invalid list option: l%c",arg[1]);
                exit(0);
                continue; /* artificial */
              case 'e': /* error line alternative */
                if (--argc <= 0)
                    noargument:			   
                emess(1,"missing argument for -%c",*arg);
                oterr = *++argv;
                continue;
              case 'W': /* specify seconds precision */
              case 'w': /* -W for constant field width */
                if ((c = arg[1]) != 0 && isdigit(c)) {
                    set_rtodms(c - '0', *arg == 'W');
                    ++arg;
                } else
                    emess(1,"-W argument missing or non-digit");
                continue;
              case 'f': /* alternate output format degrees or xy */
                if (--argc <= 0) goto noargument;
                oform = *++argv;
                continue;
              case 'r': /* reverse input */
                reversein = 1;
                continue;
              case 's': /* reverse output */
                reverseout = 1;
                continue;
              case 'd': /* set debug level */
                if (--argc <= 0) goto noargument;
                pj_ctx_set_debug( pj_get_default_ctx(), atoi(*++argv));
                continue;
              default:
                emess(1, "invalid option: -%c",*arg);
                break;
            }
            break;

        } else if (strcmp(*argv,"+to") == 0 ) {
            have_to_flag = 1;

        } else if (**argv == '+') { /* + argument */
            if( have_to_flag )
            {
                if( to_argc < MAX_PARGS )
                    to_argv[to_argc++] = *argv + 1;
                else
                    emess(1,"overflowed + argument table");
            }
            else 
            {
                if (from_argc < MAX_PARGS)
                    from_argv[from_argc++] = *argv + 1;
                else
                    emess(1,"overflowed + argument table");
            }
        } else /* assumed to be input file name(s) */
            eargv[eargc++] = *argv;
    }
    if (eargc == 0 ) /* if no specific files force sysin */
        eargv[eargc++] = "-";

    /* 
     * If the user has requested inverse, then just reverse the
     * coordinate systems.
     */
    if( inverse )
    {
        int     argcount;
        
        for( i = 0; i < MAX_PARGS; i++ )
        {
            char *arg;

            arg = from_argv[i];
            from_argv[i] = to_argv[i];
            to_argv[i] = arg;
        }

        argcount = from_argc;
        from_argc = to_argc;
        to_argc = argcount;
    }

    if( use_env_locale )
    {
        /* Set locale from environment */
        setlocale(LC_ALL, "");
    }

    if( from_argc == 0 && to_argc != 0 )
    {
        /* we will generate the from proj as the latlong of the +to in a bit */
    }
    else if (!(fromProj = pj_init(from_argc, from_argv)))
    {
        printf( "Using from definition: " );
        for( i = 0; i < from_argc; i++ )
            printf( "%s ", from_argv[i] );
        printf( "\n" );

        emess(3,"projection initialization failure\ncause: %s",
              pj_strerrno(pj_errno));
    }

    if( to_argc == 0 )
    {
        if (!(toProj = pj_latlong_from_proj( fromProj )))
        {
            printf( "Using to definition: " );
            for( i = 0; i < to_argc; i++ )
                printf( "%s ", to_argv[i] );
            printf( "\n" );
            
            emess(3,"projection initialization failure\ncause: %s",
                  pj_strerrno(pj_errno));
        }   
    }
    else if (!(toProj = pj_init(to_argc, to_argv)))
    {
        printf( "Using to definition: " );
        for( i = 0; i < to_argc; i++ )
            printf( "%s ", to_argv[i] );
        printf( "\n" );

        emess(3,"projection initialization failure\ncause: %s",
              pj_strerrno(pj_errno));
    }

    if( from_argc == 0 && toProj != NULL) 
    {
        if (!(fromProj = pj_latlong_from_proj( toProj )))
        {
            printf( "Using to definition: " );
            for( i = 0; i < to_argc; i++ )
                printf( "%s ", to_argv[i] );
            printf( "\n" );
            
            emess(3,"projection initialization failure\ncause: %s",
                  pj_strerrno(pj_errno));
        }   
    }

    if( use_env_locale )
    {
        /* Restore C locale to avoid issues in parsing/outputing numbers*/
        setlocale(LC_ALL, "C");
    }

    if (mon) {
        printf( "%c ---- From Coordinate System ----\n", tag );
        pj_pr_list(fromProj);
        printf( "%c ---- To Coordinate System ----\n", tag );
        pj_pr_list(toProj);
    }

    /* set input formating control */
    if( !fromProj->is_latlong )
        informat = strtod;
    else {
        informat = dmstor;
    }

    if( !toProj->is_latlong && !oform )
        oform = "%.2f";

    /* process input file list */
    for ( ; eargc-- ; ++eargv) {
        if (**eargv == '-') {
            fid = stdin;
            emess_dat.File_name = "<stdin>";

        } else {
            if ((fid = fopen(*eargv, "rt")) == NULL) {
                emess(-2, *eargv, "input file");
                continue;
            }
            emess_dat.File_name = *eargv;
        }
        emess_dat.File_line = 0;
        process(fid);
        fclose(fid);
        emess_dat.File_name = 0;
    }

    if( fromProj != NULL )
        pj_free( fromProj );
    if( toProj != NULL )
        pj_free( toProj );

    pj_deallocate_grids();

    exit(0); /* normal completion */
}
